File: ios/Objective-C/GeneralSettingsObjC/GeneralSettingsObjC/Modules/ViewSettings/ViewSettingsViewController.m
//
//  ViewSettingsViewController.m
//  GeneralSettings
//
//  Created by dynamsoft on 2021/11/19.
//

#import "ViewSettingsViewController.h"

static NSString *displayOverlayTag      = @"100";
static NSString *displayTorchButtonTag      = @"101";

@interface ViewSettingsViewController ()<UITableViewDelegate, UITableViewDataSource>
{
    NSArray *viewSettingDataArray;
    NSMutableDictionary *recordViewSettingStateDic;
}
@property (nonatomic, strong) UITableView *viewSettingsTableView;

@end

@implementation ViewSettingsViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.view.backgroundColor = [UIColor whiteColor];
    
    self.title = @"View Settings";
    
    viewSettingDataArray = @[[GeneralSettingsHandle setting].dceViewSettings.displayOverlay,
                             [GeneralSettingsHandle setting].dceViewSettings.displayTorchButton
    ];
    
    recordViewSettingStateDic = [NSMutableDictionary dictionary];
    if ([GeneralSettingsHandle setting].dceViewSettings.displayOverlayIsOpen) {
        [recordViewSettingStateDic setValue:@(1) forKey:[GeneralSettingsHandle setting].dceViewSettings.displayOverlay];
    } else {
        [recordViewSettingStateDic setValue:@(0) forKey:[GeneralSettingsHandle setting].dceViewSettings.displayOverlay];
    }
    
    if ([GeneralSettingsHandle setting].dceViewSettings.displayTorchButtonIsOpen) {
        [recordViewSettingStateDic setValue:@(1) forKey:[GeneralSettingsHandle setting].dceViewSettings.displayTorchButton];
    } else {
        [recordViewSettingStateDic setValue:@(0) forKey:[GeneralSettingsHandle setting].dceViewSettings.displayTorchButton];
    }
    
    [self setupUI];
}

- (void)setupUI
{
    self.viewSettingsTableView = [[UITableView alloc] initWithFrame:self.view.bounds style:UITableViewStylePlain];
    self.viewSettingsTableView.backgroundColor = [UIColor whiteColor];
    self.viewSettingsTableView.delegate = self;
    self.viewSettingsTableView.dataSource = self;
    self.viewSettingsTableView.separatorStyle = UITableViewCellSeparatorStyleNone;
    [self.view addSubview:self.viewSettingsTableView];
}


//MARK: UITableViewDelegate
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return viewSettingDataArray.count;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return [BasicTableViewCell cellHeight];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSString *dceViewSettingString = viewSettingDataArray[indexPath.row];
    NSInteger switchState = [[recordViewSettingStateDic valueForKey:dceViewSettingString] integerValue];
    
    weakSelfs(self)
    static NSString *identifier = @"basicSwitchCellIdentifier";
    BasicSwitchTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
    if (!cell) {
        cell = [[BasicSwitchTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:identifier];
    }
    
    cell.selectionStyle = UITableViewCellSelectionStyleNone;
    
    
    if ([dceViewSettingString isEqualToString:[GeneralSettingsHandle setting].dceViewSettings.displayOverlay]) {
        cell.questionButton.hidden = NO;
        cell.questionBlock = ^{
            [[ToolsHandle toolManger] addAlertViewWithTitle:@"Display Overlay" Content:displayOverlayExplain actionTitle:nil ToView:weakSelf completion:^{
                            
            }];
        };
    } else {
        cell.questionButton.hidden = YES;
    }
    
    cell.switchChangedBlock = ^(BOOL isOn) {
        [weakSelf handleSwitchWithDCESettingString:dceViewSettingString andState:isOn];
    };

    [cell updateUIWithTitle:dceViewSettingString withSwitchState:switchState];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
  
}

///Make configurations on the camera view.
- (void)handleSwitchWithDCESettingString:(NSString *)dceViewSettingString andState:(BOOL)isOn
{
    DCEViewSettings dceViewSettings = [GeneralSettingsHandle setting].dceViewSettings;
    
    if ([dceViewSettingString isEqualToString:[GeneralSettingsHandle setting].dceViewSettings.displayOverlay]) {

        // Make configurations for overlays.
        // Highlighted overlays will be displayed on the successfully decoded barcodes when overlayVisible is set to true.
        if (isOn) {
            dceViewSettings.displayOverlayIsOpen = YES;
            [GeneralSettingsHandle setting].dceViewSettings = dceViewSettings;
            
            [GeneralSettingsHandle setting].cameraView.overlayVisible = true;
        } else {
            dceViewSettings.displayOverlayIsOpen = NO;
            [GeneralSettingsHandle setting].dceViewSettings = dceViewSettings;
            
            [GeneralSettingsHandle setting].cameraView.overlayVisible = false;
        }
        
    } else if ([dceViewSettingString isEqualToString:[GeneralSettingsHandle setting].dceViewSettings.displayTorchButton]) {

        // Make configurations for overlays.
        // Setting the torchButtonVisible to true will display a torch button on the UI.
        // The torch button can control the status of the mobile torch.
        if (isOn) {
            dceViewSettings.displayTorchButtonIsOpen = YES;
            [GeneralSettingsHandle setting].dceViewSettings = dceViewSettings;
            
            [GeneralSettingsHandle setting].cameraView.torchButtonVisible = true;
        } else {
            dceViewSettings.displayTorchButtonIsOpen = NO;
            [GeneralSettingsHandle setting].dceViewSettings = dceViewSettings;
            
            [GeneralSettingsHandle setting].cameraView.torchButtonVisible = false;
        }
        
        
    }
}




@end

File: ios/Swift/GeneralSettingsSwift/GeneralSettingsSwift/ViewControllers/ViewSettings/ViewSettingsViewController.swift
//
//  ViewSettingsViewController.swift
//  GeneralSettingsSwift
//
//  Copyright © Dynamsoft. All rights reserved.
//

import UIKit

class ViewSettingsViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {

    private var viewSettingsDataArray: [String] = []
    
    /// Record view-settings switch state dic.
    private var recordViewSettingsSwitchStateDic: [String : Bool] = [:]
    
    private lazy var viewSettingsTableView = {
        let tableView = UITableView.init(frame: self.view.bounds, style: .plain)
        tableView.backgroundColor = .white
        tableView.delegate = self
        tableView.dataSource = self
        tableView.separatorStyle = .none
        return tableView
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        self.view.backgroundColor = .white
        self.title = "View Settings";
        
        handleData()
        setupUI()
    }

    private func handleData() -> Void {
        viewSettingsDataArray.removeAll()
        recordViewSettingsSwitchStateDic.removeAll()
        
        let basicDataArray: [String] = [GeneralSettings.shared.dceViewSettings.displayOverlay,
                                        GeneralSettings.shared.dceViewSettings.displayTorchButton
        ]
        viewSettingsDataArray.append(contentsOf: basicDataArray)
        
        if GeneralSettings.shared.dceViewSettings.displayOverlayIsOpen == true {
            recordViewSettingsSwitchStateDic[GeneralSettings.shared.dceViewSettings.displayOverlay] = true
        } else {
            recordViewSettingsSwitchStateDic[GeneralSettings.shared.dceViewSettings.displayOverlay] = false
        }
        
        if GeneralSettings.shared.dceViewSettings.displayTorchButtonIsOpen == true {
            recordViewSettingsSwitchStateDic[GeneralSettings.shared.dceViewSettings.displayTorchButton] = true
        } else {
            recordViewSettingsSwitchStateDic[GeneralSettings.shared.dceViewSettings.displayTorchButton] = false
        }
    }
    
    private func setupUI() -> Void {
        self.view.addSubview(viewSettingsTableView)
    }
    
    // MARK: - UITableViewDelegate
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return BasicTableViewCell.cellHeight()
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.viewSettingsDataArray.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let titleTag = self.viewSettingsDataArray[indexPath.row]
        
        let identifier = "BasicSwitchTableViewCell"
        var cell = tableView.dequeueReusableCell(withIdentifier: identifier) as? BasicSwitchTableViewCell
        if cell == nil {
            cell = BasicSwitchTableViewCell.init(style: .default, reuseIdentifier: identifier)
        }
        
        if titleTag == GeneralSettings.shared.dceViewSettings.displayOverlay {
            cell?.questionButton.isHidden = false
        } else {
            cell?.questionButton.isHidden = true
        }
        
        cell?.questionCompletion = {
            [unowned self] in
            self.handleDCEExplain(with: indexPath, settingString: titleTag)
        }
        
        cell?.switchChangedCompletion = {
            [unowned self] isOn in
            self.handleViewSettingSwitch(with: indexPath, settingString: titleTag, isOn: isOn)
        }
        
        cell?.updateUI(with: titleTag, isOn:recordViewSettingsSwitchStateDic[titleTag]!)
        return cell!
    }
    
    // MARK: - HandleOperation
    
    private func handleViewSettingSwitch(with indexPath: IndexPath,
                                         settingString: String,
                                         isOn: Bool) -> Void {
        if settingString == GeneralSettings.shared.dceViewSettings.displayOverlay {
            
            // Make configurations for overlays.
            // Highlighted overlays will be displayed on the successfully decoded barcodes when overlayVisible is set to true.
            GeneralSettings.shared.cameraView.overlayVisible = isOn
            GeneralSettings.shared.dceViewSettings.displayOverlayIsOpen = isOn
        } else if settingString == GeneralSettings.shared.dceViewSettings.displayTorchButton {
            
            // Make configurations for overlays.
            // Setting the torchButtonVisible to true will display a torch button on the UI.
            // The torch button can control the status of the mobile torch.
            GeneralSettings.shared.cameraView.torchButtonVisible = isOn
            GeneralSettings.shared.dceViewSettings.displayTorchButtonIsOpen = isOn
        }
    }
    
    private func handleDCEExplain(with indexPath: IndexPath,
                                  settingString: String) -> Void {
        if settingString == GeneralSettings.shared.dceViewSettings.displayOverlay {
            ToolsManager.shared.addAlertView(to: self,title: settingString, content: displayOverlayExplain, completion: nil)
        }
    }
}
